Refactor modal_aero_amicphys_intr

huiwanpnnl opened the pull request:
`modal_aero_amicphys_intr` is the top-level driver subroutine of MAM's aerosol microphysics package. It does 4 things:

1. Constructs cloudy and clear (cloud-free) subareas within a grid cell using grid-cell mean info from the host model,
2. Calls `mam_amicphys_1gridcell` to calculate the microphysics and update the subarea mixing ratios,
3. Use the updated subarea mixing ratios to form grid cell mean values.
4. Diagnose some tendencies for output.

The original codes for 1, 2, and 4 consisted of a lot of do-loops and if-blocks, some of which were not so straightforward to understand. For this PR, I have done the following
- Reworked all the do-loops and if-blocks in steps 1, 2, and 4 and created subroutines. The new subroutines are in new files `modal_aero_amicphys_subareas.F90` and `modal_aero_amicphys_diags.F90`.
- Added comments to explain the overall flow of the calculation.
- Moved `use modal_aero_amicphys_control` from the top of the module to inside each subroutine; added `only` in subroutines that have already been refactored.
- Removed unused variables.
- Removed all codes contained in `#if ( defined( MOSAIC_SPECIES ) )...#endif`.
- Removed all codes contained in `#if ( defined( CAMBOX_ACTIVATE_THIS ) )...#endif` that were used only for writing info to log files.

There are quite a few variables named like `qgcm1`, `qgcm2`, `qgcm3`, `qgcm4` etc.. The numbers correspond to different places in the time integration loop, and these different values are used for coupling the microphysics processes in a specific way. It might be useful to change the numbers to more descriptive suffixes. I will think about that as I refactor the remaining interface subroutines in `modal_aero_amicphys.F90`.

mingxuanwupnnl committed:
Refactor nucleate_ice_cam_calc

mingxuanwupnnl committed:
Remove unused code for contact freezing, bam aerosol in microp_aero.F90

mingxuanwupnnl committed:
Further remove unused variables

mingxuanwupnnl committed:
Add comments for units

huiwanpnnl committed:
amicphys subareas: separate code into a tmp file.

huiwanpnnl committed:
amicphys subareas: separate code into a tmp file.

huiwanpnnl committed:
amicphys subareas: new subr. setup_subareas

huiwanpnnl committed:
amicphys subareas: new subr. setup_subareas

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_relhum

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_relhum

huiwanpnnl committed:
amicphys subareas: new subr. copy_gcm_to_subarea

huiwanpnnl committed:
amicphys subareas: new subr. copy_gcm_to_subarea

huiwanpnnl committed:
amicphys subareas: use subr. copy_gcm_to_subarea for gases

huiwanpnnl committed:
amicphys subareas: use subr. copy_gcm_to_subarea for gases

huiwanpnnl committed:
amicphys subareas: some clean-up and cosmetic changes

huiwanpnnl committed:
amicphys subareas: some clean-up and cosmetic changes

huiwanpnnl committed:
amicphys subareas: update some comments

huiwanpnnl committed:
amicphys subareas: update some comments

huiwanpnnl committed:
amicphys subareas: unify code for cloudy-only and clear-only cases

huiwanpnnl committed:
amicphys subareas: unify code for cloudy-only and clear-only cases

huiwanpnnl committed:
amicphys subareas: revise overall structure of code block

huiwanpnnl committed:
amicphys subareas: revise overall structure of code block

huiwanpnnl committed:
amicphys subareas: rename subr. copy_gcm_to_subarea to copy_cnst

huiwanpnnl committed:
amicphys subareas: rename subr. copy_gcm_to_subarea to copy_cnst

huiwanpnnl committed:
amicphys subareas: separate partitioning of cloud-borne aerosols

huiwanpnnl committed:
amicphys subareas: separate partitioning of cloud-borne aerosols

huiwanpnnl committed:
amicphys subareas: new subr.'s for partitioning cldbrn aerosols

huiwanpnnl committed:
amicphys subareas: new subr.'s for partitioning cldbrn aerosols

huiwanpnnl committed:
amicphys subareas: new subr. get_partition_factors

huiwanpnnl committed:
amicphys subareas: new subr. get_partition_factors

huiwanpnnl committed:
amicphys subareas: split loop for the partitioning of mass and number

huiwanpnnl committed:
amicphys subareas: split loop for the partitioning of mass and number

huiwanpnnl committed:
amicphys subareas: a bugfix and some comments

huiwanpnnl committed:
amicphys subareas: a bugfix and some comments

huiwanpnnl committed:
amicphys subareas: further revise loops

huiwanpnnl committed:
amicphys subareas: further revise loops

huiwanpnnl committed:
amicphys subareas: rename loop/array indices

huiwanpnnl committed:
amicphys subareas: rename loop/array indices

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_q_numb_for_intrst_aerosols

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_q_numb_for_intrst_aerosols

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_q_mass_for_intrst_aerosols

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_q_mass_for_intrst_aerosols

huiwanpnnl committed:
amicphys subareas: small mods in dummy arguments

huiwanpnnl committed:
amicphys subareas: small mods in dummy arguments

huiwanpnnl committed:
amicphys subareas: new subr. for gas mixing rations in clear subarea

huiwanpnnl committed:
amicphys subareas: new subr. for gas mixing rations in clear subarea

huiwanpnnl committed:
amicphys subareas: improve comments

huiwanpnnl committed:
amicphys subareas: improve comments

huiwanpnnl committed:
amicphys subareas: revise set_subarea_q_numb_for_intrst_aerosols

...so that it partitions only one state.

huiwanpnnl committed:
amicphys subareas: revise set_subarea_q_numb_for_intrst_aerosols

...so that it partitions only one state.

huiwanpnnl committed:
amicphys subareas: revise set_subarea_q_mass_for_intrst_aerosols

...so that it partitions only one state.

huiwanpnnl committed:
amicphys subareas: revise set_subarea_q_mass_for_intrst_aerosols

...so that it partitions only one state.

huiwanpnnl committed:
amicphys subareas: revise subroutine names

huiwanpnnl committed:
amicphys subareas: revise subroutine names

huiwanpnnl committed:
amicphys subareas: prep for new subr.

huiwanpnnl committed:
amicphys subareas: prep for new subr.

huiwanpnnl committed:
amicphys subareas: cosmetic changes

huiwanpnnl committed:
amicphys subareas: cosmetic changes

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_gases_and_aerosols

huiwanpnnl committed:
amicphys subareas: new subr. set_subarea_gases_and_aerosols

huiwanpnnl committed:
amicphys subareas: improve variable names and comments

huiwanpnnl committed:
amicphys subareas: improve variable names and comments

huiwanpnnl committed:
amicphys subreas to gcm: split and simplify loops

huiwanpnnl committed:
amicphys subreas to gcm: split and simplify loops

huiwanpnnl committed:
amicphys subareas: new subr. get_gcm_gases_and_aerosols_from_subares

huiwanpnnl committed:
amicphys subareas: new subr. get_gcm_gases_and_aerosols_from_subares

huiwanpnnl committed:
amicphys subarea to gcm: done

huiwanpnnl committed:
amicphys subarea to gcm: done

huiwanpnnl committed:
add 'only' to 'use modal_aero_amicphys_control' in some subroutines

huiwanpnnl committed:
add 'only' to 'use modal_aero_amicphys_control' in some subroutines

huiwanpnnl committed:
amicphys_intr: remove some unused code

Most are related to MOSAIC_SPECIES

huiwanpnnl committed:
amicphys_intr: remove some unused code

Most are related to MOSAIC_SPECIES

huiwanpnnl committed:
amicphys_intr: remove some unused variables

huiwanpnnl committed:
amicphys_intr: remove some unused variables

huiwanpnnl committed:
amicphys_intr: rename loop indices i,k,n,l => ii,kk,imode,icnst

huiwanpnnl committed:
amicphys_intr: rename loop indices i,k,n,l => ii,kk,imode,icnst

huiwanpnnl committed:
amicphys_intr: remove unused contents

huiwanpnnl committed:
amicphys_intr: remove unused contents

huiwanpnnl committed:
amicphys_intr: new subr. amicphys_diags_init

huiwanpnnl committed:
amicphys_intr: new subr. amicphys_diags_init

huiwanpnnl committed:
amicphys_intr: new subr. get_gcm_tend_diags_from_subareas

huiwanpnnl committed:
amicphys_intr: new subr. get_gcm_tend_diags_from_subareas

huiwanpnnl committed:
amicphys_intr: new subr. accumulate_column_tend_integrals

huiwanpnnl committed:
amicphys_intr: new subr. accumulate_column_tend_integrals

huiwanpnnl committed:
amicphys_intr: new subr. outfld_1proc_all_cnst

huiwanpnnl committed:
amicphys_intr: new subr. outfld_1proc_all_cnst

mingxuanwupnnl committed:
Address PR review comments, mainly move pbuf out of nucleate_ice_cam_calc

mingxuanwupnnl committed:
Change t to temperature in microp_aero_run, change some variables in nucleate_ice.F90

huiwanpnnl committed:
amicphys_intr: work on comments, units, etc.

huiwanpnnl committed:
amicphys_intr: work on comments, units, etc.

huiwanpnnl committed:
amicphys_intr: refactor a small loop

huiwanpnnl committed:
amicphys_intr: refactor a small loop

mingxuanwupnnl committed:
use common min_max function, and replace i, k with icol and kk

mingxuanwupnnl committed:
Change the comments for pbuf variables

overfelt commented:
@huiwanpnnl , I have been looking at some of the initialization code to figure out how the mam_refactor tests initialize the box model and how this fits into mam4xx.   I'm not sure how much of the initialization code is needed for testing in mam4xx and where the Fortran code that is to be converted to C++ should be implemented in mam4xx.  Maybe we could get together with @jeff-cohere and you could educate us?

jeff-cohere commented:
I haven't looked at this yet, but if I had to guess, I'd say this logic is likely related to the SCREAM/EAMxx atmosphere process that interacts with the Atmosphere Driver. That said, I wouldn't mind a get-together to go over some of the basics here.

jeff-cohere commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,649 @@
+  real(wp), intent(out) :: fclea, fcldy               ! area fraction of clear/cloudy subarea [unitless]
+
+  ! Cloud chemistry is only active when cld(i,k) >= 1.0e-5_wp
+  ! It may be that the macrophysics has a higher threshold that this
```
that -> than

huiwanpnnl replied:
Corrected.

jeff-cohere commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,649 @@
+
+     call copy_cnst( qsub2(:,jclea), qsub3(:,jclea), cnst_is_gas ) !from, to, flag
+
+     ! Calculater the gas mixing ratios in the cloudy subarea using the grid-cell mean, 
```
Calculater -> Calculate

huiwanpnnl replied:
Corrected.

jeff-cohere commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,649 @@
+
+end subroutine set_subarea_gases_and_aerosols
+
+subroutine copy_cnst( q_in, q_copy, lcopy )
```
@huiwanpnnl , do you think this subroutine is more clear than the `where` clause that does the copy itself? Are people confused by the use of `where`? To me, this is a simple construction that I wish other programming languages had.

huiwanpnnl replied:
Ah, I see what you mean.

This subroutine `copy_cnst` is a remnant of several rounds of revision. Earlier, it had a longer name and several additional input arguments. Now that it has become this simple, we can simple write `where (lcopy) q_copy = q_in` in the calling routine. I removed the `copy_cnst` subroutine in [1b769fd](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/1b769fdf84fcb3e4f395d22cd06225b4b54b1f98).

jeff-cohere replied:
Thanks for the historical perspective, Hui!

huiwanpnnl commented:
> @huiwanpnnl , I have been looking at some of the initialization code to figure out how the mam_refactor tests initialize the box model and how this fits into mam4xx. I'm not sure how much of the initialization code is needed for testing in mam4xx and where the Fortran code that is to be converted to C++ should be implemented in mam4xx. Maybe we could get together with @jeff-cohere and you could educate us?

Hi @overfelt and @jeff-cohere, in the box model (`mam_refactor`), I refactored the microphysics initialization subroutine `modal_aero_amicphys_init` in `modal_aero_amicphys.F90` a while ago. It seems the aerosol microphysics code assumes MAM's internal "bookkeeping" for the tracers/constituents may be different from the host model's, so we establish a mapping between the two systems, copy some information from the host model, and then do a bunch of initialization for the microphysics package's internal bookkeeping. I have not spent a lot of time studying the mapping part yet.

MAMxx also will contain parts that are not aerosol microphysics (i.e., emissions, dry removal, interactions with clouds,  etc.). My understanding is those parts use the host models `state%q` array and `pbuf`, but the colleagues who are refactoring those codes will know better :-)

In any case, I'd be happy to get together and have a discussion. It will be useful to include Balwinder and the others (Mingxuan, Shuaiqi), too.

huiwanpnnl commented:
Also, I guess we should take the initialization discussion to a different place - perhaps a separate "issue"?

jeff-cohere commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,649 @@
+
+  ! Calculate subarea-mean mixing ratios
+
+  tmp_q_cldbrn_cldy = qgcm_cldbrn/fcldy                                              ! cloud-borne,  cloudy subarea
```
Does this assume that `fcldy` is positive? If so, we should document that assumption.

huiwanpnnl replied:
Yes. I've added the following comment:

```
  ! Note that this subroutine is designed for partially cloudy grid cells,
  ! hence both fclea and fcldy are assumed to be nonzero.
```

jeff-cohere commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,649 @@
+  tmp_q_cldbrn_cldy = qgcm_cldbrn/fcldy                                              ! cloud-borne,  cloudy subarea
+  tmp_q_intrst_cldy = max( 0.0_wp, ((qgcm_intrst+qgcm_cldbrn) - tmp_q_cldbrn_cldy) ) ! interstitial, cloudy subarea
+
+  tmp_q_intrst_clea = (qgcm_intrst - fcldy*tmp_q_intrst_cldy)/fclea                  ! interstitial, clear  subarea
```
There is a similar assumption about `fclea` being positive here.

huiwanpnnl replied:
Same as above. Added the following comment:

```
  ! Note that this subroutine is designed for partially cloudy grid cells,
  ! hence both fclea and fcldy are assumed to be nonzero.
```

huiwanpnnl committed:
amicphys_intr: rm subr. copy_cnst; correct typos

huiwanpnnl committed:
amicphys_intr: rm subr. copy_cnst; correct typos

huiwanpnnl committed:
subr. get_partition_factors: document assumption of flea/=0 and fcldy/=0

huiwanpnnl committed:
subr. get_partition_factors: document assumption of flea/=0 and fcldy/=0

singhbalwinder committed:
Merge pull request #43 from eagles-project/mingxuanwupnnl/atm/refactor_nucleate_ice_cam_calc

Refactor subroutine nucleate_ice_cam_calc in nucleate_ice_cam.F90

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_diags.F90:
```diff
@@ -0,0 +1,168 @@
+
+  implicit none
+
+  public
```
Fortran modules are by default public, It is better to make everything private and then explicitly declare the public data and member functions/subroutines.

huiwanpnnl replied:
Good point! Addressed in [7e41ef7](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/7e41ef7f96340788efc08fb9abf092839d82c223)

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_diags.F90:
```diff
@@ -0,0 +1,168 @@
+
+  use modal_aero_amicphys_control, only: do_q_coltendaa, do_qqcw_coltendaa, &
+                                         iqtend_cond, iqtend_rnam, iqtend_nnuc, iqtend_coag, &
+                                         iqqcwtend_rnam
```
We have a lot of variables in `modal_aero_amicphys_control`. For these variables, would it be possible to remove them from `modal_aero_amicphys_contro` and declare them here as `protected`? It will spread out variables and also guarantees that they won't be changed anywhere outside of this module.

huiwanpnnl replied:
Done as suggested in [a215508](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/a2155082dbc19f3c00f4fdc564461665a05cfe18)+[c5488b7](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/c5488b7cfeeaac0d3b5aae0334795f2353f6174b). (I also moved the subroutine `m_a_amicphys_init_history` in order to be able to declare some of the variables as `protected`.)

huiwanpnnl replied:
BTW, @singhbalwinder, we might want to meet up and discuss a topic loosely related to your comment above: In the box model, I have moved some module variables from `modal_aero_amicphys_control` to new modules named `modal_aero_microp_modes`, `modal_aero_microp_species`, and `modal_aero_microp_host_mapping`. Do you want those changes to be ported to this repo?

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_diags.F90:
```diff
@@ -0,0 +1,168 @@
+                                             qsub_tendaa, qqcwsub_tendaa,       &! in
+                                             qgcm_tendaa, qqcwgcm_tendaa        )! out
+
+  use modal_aero_amicphys_control, only: wp=>r8, ncnst=>gas_pcnst, maxsubarea, nqtendaa, nqqcwtendaa
```
I see a lot of `aa` at the end of the variable names. Do you know what it stands for?

huiwanpnnl replied:
I don't know for sure. My impression is that `aa` is used for distinguishing those variables from another set of variables ending in `bb`. Those `bb` variables seems to be used only in the box model. Both the `aa` and `bb` variables are used for diagnostics only. 

I am guessing that, at some point, somebody will need to decide how to organize the diagnostics in MAMXX. At that point, it would be useful to organize those `aa` variables in to a data structure or replace `aa` with `_diag` or something similar.

A related thought: I am guessing the BFB testing mechanism we are using for the Fortran code refactoring does not cover the aerosol diagnostics (i.e., those calculations that are used only when `history_aerosol = .t.`, `history_verbose = .t.` etc. and do not affect the time integration). Is that correct? If so, we would want to keep in mind that at a later point, those diagnostics need to be revisited.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,637 @@
+
+  implicit none
+
+  public
```
Same here, it is better to make it private and state the public members explicitly.

huiwanpnnl replied:
Addressed in [3d1b1b6](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/3d1b1b6daa390de0850fe74966fe2c3bfd53afd2).

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys.F90:
We can use the `min_max_bound` function here (in chemistry/utils/mam_support.F).

huiwanpnnl replied:
Done in [a450023](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/a450023ce43b854831ae26eadfef02d02d5ffa01) after rebasing branch onto `refactor-maint-2.0` in [0e94531](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/0e9453145d441bbe4e1f190512a27d0cb7deb6d9).

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,637 @@
+#if ( defined( MAM_STANDALONE ) )
+  else if (cldy_rh_sameas_clear > 0) then
+     relhumsub(:) = relhumgcm
+#endif
```
Would it be okay to simply remove these codes between the directives? Do you see any reason to keep these codes? Is it being used in the box model? If yes, we can write a note here for the C++ port to let  the team not to port this code.

huiwanpnnl replied:
These are indeed used in the box model, but I've removed them - and also removed similar blocks in `modal_aero_amicphys.F90` in [6a76cee](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/6a76cee38eba488e62680a070cff0c8c712e75db). Porting the refactored code to the box model will require some further code review and comparison, which I will do sometime in the next months.

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,637 @@
+
+     if (jclea > 0) then
+        relhum_tmp = (relhumgcm - afracsub(jcldy))/afracsub(jclea)
+        relhumsub(jclea) = max( 0.0_wp, min( 1.0_wp, relhum_tmp ) )
```
We can use `min_max_bound` here as well.

huiwanpnnl replied:
Done in [a450023](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/a450023ce43b854831ae26eadfef02d02d5ffa01) after rebasing branch onto `refactor-maint-2.0` in [0e94531](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/0e9453145d441bbe4e1f190512a27d0cb7deb6d9).

singhbalwinder commented on components/eam/src/chemistry/modal_aero/modal_aero_amicphys_subareas.F90:
```diff
@@ -0,0 +1,637 @@
+  !---------------------------------------------------------------------------------------------------
+  grid_cell_has_only_clea_area = ((jclea == 1) .and. (jcldy == 0) .and. (nsubarea == 1))
+  grid_cell_has_only_cldy_area = ((jclea == 0) .and. (jcldy == 1) .and. (nsubarea == 1))
+  gird_cell_is_partly_cldy = (jclea > 0) .and. (jcldy > 0) .and.  (jclea+jcldy == 3) .and. (nsubarea == 2)
```
It's a great way to use meaningful logical variables!

huiwanpnnl replied:
Thanks :-)

singhbalwinder reviewed (approved):
Thanks, Hui for your hard work on this! It came out really good! It is much easier to understand the flow of code. I have left some minor comments. There are some single letter variables (e.g., `t` for temperature) which should be modified if their scope is more than 4-5 lines of the code. This can be done in this PR or the next one.

huiwanpnnl committed:
make all entities private by default in modal_aero_amicphys_diags.F90

huiwanpnnl committed:
make all entities private by default in modal_aero_amicphys_diags.F90

huiwanpnnl committed:
move some variables from modal_aero_amicphys_control.F90 to modal_aero_amicphys_diags.F90

huiwanpnnl committed:
move some variables from modal_aero_amicphys_control.F90 to modal_aero_amicphys_diags.F90

huiwanpnnl committed:
move subr. m_a_amicphys_init_history

...from modal_aero_amicphys.F90 to modal_aero_amicphys_diags.F90
so that some variables in the latter can be declared as protected.

huiwanpnnl committed:
move subr. m_a_amicphys_init_history

...from modal_aero_amicphys.F90 to modal_aero_amicphys_diags.F90
so that some variables in the latter can be declared as protected.

huiwanpnnl committed:
declare entities to be private by default in modal_aero_amicphys_subareas.F90

huiwanpnnl committed:
declare entities to be private by default in modal_aero_amicphys_subareas.F90

huiwanpnnl committed:
rename variables t, q to temp, qq in modal_aero_amicphys.F90

huiwanpnnl committed:
rename variables t, q to temp, qq in modal_aero_amicphys.F90

huiwanpnnl commented:
Single-letter variables `t` and `q` are renamed to `temp` and `qq` in [e347b68](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/e347b68549ffa7b496afcda87542a4568607ae09).

huiwanpnnl committed:
remove lines used only by the box model

huiwanpnnl committed:
remove lines used only by the box model

huiwanpnnl committed:
rebase branch on to refactor-maint-2.0

Merge branch 'huiwanpnnl/refactor-amicphys-intr' of github.com:eagles-project/e3sm_mam4_refactor
into local branch huiwanpnnl/refactor-amicphys-intr

huiwanpnnl committed:
use function min_max_bound

huiwanpnnl commented:
Note that above, I rebased my branch onto `refactor-maint-2.0` in order to be able to use `min_max_bound`, and then I did a `git pull` and `git push`. This has resulted in the commit history on my branch becoming rather confusing. To resolve this problem, the following was done: 
- create a new branch [`huiwanpnnl/refactor-amicphys-intr2`](https://github.com/eagles-project/e3sm_mam4_refactor/tree/huiwanpnnl/refactor-amicphys-intr2) based on this branch,
- rewind back to [6a76cee](https://github.com/eagles-project/e3sm_mam4_refactor/pull/49/commits/6a76cee38eba488e62680a070cff0c8c712e75db), 
- redo the rebase, 
- commit the changes that use function `min_max_bound`,
- push the new branch to origin.

The new branch has a clean history. I therefore close this PR and issue a new one (#53).
